src/compile/data/source.ts

```diff
@@ -20,35 +20,37 @@ export class SourceNode extends DataFlowNode {
     data = data || {name: 'source'};
 
     if (isInlineData(data)) {
-      this._data = data;
+      this._data = {values: data.values};
     } else if (isUrlData(data)) {
-      // Extract extension from URL using snippet from
-      // http://stackoverflow.com/questions/680929/how-to-extract-extension-from-filename-string-in-javascript
-      let defaultExtension = /(?:\.([^.]+))?$/.exec(data.url)[1];
-      if (!contains(['json', 'csv', 'tsv', 'topojson'], defaultExtension)) {
-        defaultExtension = 'json';
+      this._data = {url: data.url};
+
+      if (!data.format || !data.format.type) {
+        // Extract extension from URL using snippet from
+        // http://stackoverflow.com/questions/680929/how-to-extract-extension-from-filename-string-in-javascript
+        let defaultExtension = /(?:\.([^.]+))?$/.exec(data.url)[1];
+        if (!contains(['json', 'csv', 'tsv', 'topojson'], defaultExtension)) {
+          defaultExtension = 'json';
+        }
+
+        if (defaultExtension !== 'json') {
+          if (!data.format) {
+            data.format = {};
+          }
+          data.format.type = defaultExtension as DataFormatType;
```

Comment:
Why do we need `as` here?  Please add a comment to explain.

Reply:
defaultExtension has type string but we ensure that it is DataFormatType above

done